src/rustc_lexer/src/cursor.rs

```diff
@@ -0,0 +1,40 @@
+use std::str::Chars;
+
+pub(crate) struct Cursor<'a> {
+    chars: Chars<'a>,
+    prev: char,
+}
+
+pub(crate) const EOF_CHAR: char = '\0';
```

Comment:
You should use `Option<char>` nowadays, as it's the same size as `char` (`None` uses one of the values outside of the 21-bit Unicode range), and [`c == Some('x')` optimizes to the same thing as `c == 'x'`](https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=5b0e96e82968aa6759ce81f8d72e608b).

Reply:
Matches against option would look uglier in source code: adding Some everywhere would mostly add noise, because all matches are with _ branch anyway. But yeah, returning option here would certainly be more mundane

Reply:
Oh, I see, I would've expected code to first unpack it (maybe with `?`) in most situations, eagerly handling EOF.

Reply:
The thing is, handling EOF and handling random character not continuing the current token is almost always the same, so handling EOF specially will lead to micro code duplication. If you handle EOF via `?`, you should handle "other" chars via explicit `return None`.

Anyway, I have not strong opinion here, could flip to Option if you believe that's better.

Reply:
Ah, I guess I might need to wait for everything else to be in the final form before I can take a closer look at EOF, don't mind me until then!